### ISSUE TITLE
new check : statement lambda can be replaced with expression lambda

### ISSUE THREAD

**Issue by Dichotomia** [checkstyle/checkstyle#6850]
Similar to Intellij Inspection.

Description : Reports lambda expressions with code block bodies when expression-style bodies can be used

**Comment by strkkk** [comment]
@Dichotomia please provide code with violations and config example.

**Comment by Dichotomia** [comment]
> 
> 
> @Dichotomia please provide code with violations and config example.

I will do that as soon as possible

**Comment by Dichotomia** [comment]
```
/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
       <module name="UnnecessaryLambdaBlock"/>
    </module>
</module>

/var/tmp $ cat YOUR_FILE.java
	public class Test {

		Runnable oneLineValid = () -> System.out.println("Hello world"); // No Violation

		Runnable oneLineInvalid = () -> { // Violation
			System.out.println("Hello world");
		};

		Runnable longValid = () -> { // No Violation
			System.out.println("Hello world");
			System.out.println("Hello world2");
		};

	}

**Comment by romani** [comment]
in general intention is good.
BUT lambda with block (`{ .... }`) and as expression have one more parameter implied - limit of line length for code. One more aspect is as soon lambda become multiline, is it not a rare demand to enforce `{` around.

Please share your ideas/experience on behavior for cases like (imagine):
`Runnable oneLineValid = () -> System.out.println("Heeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeello"); // ????`
`Runnable oneLineValid = () -> {
  System.out.println("Heeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeello")} // ????
; `

Do want to have a Check only for no argument lambda ? What is expected behavior on all possible type and forms of lambda? Please share more examples and try to think about lambdas that a part of chain calls .

If Check is approved, we would have to apply it to our code, you can review our code, search for "->" and propose a change to code.
Example of non simple lambda:
https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/EqualsHashCodeCheck.java#L163-L166

**Comment by Dichotomia** [comment]
For the examples you give:

Runnable oneLineValid = () -> System.out.println("Heeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeello"); // No violation

Runnable oneLineValid = () -> {
  System.out.println("Heeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeello")} // It depend of the following lines
;

**Comment by romani** [comment]
> It depends of the following lines

Please explain, I lost your thought.

>lambda is not advised because filter is not 

Please suggest logic on how to skip such cases. If that is not possible to do without false-positives, unfortunately I can not approve such Check in main repo, but we can host it at sevntu-checks project (our sandbox project)

**Comment by Dichotomia** [comment]
> 
> 
> > It depends of the following lines
> 
> Please explain, I lost your thought.

It depends if there is an other instruction after that instruction.
Runnable oneLineValid = () -> {
System.out.println("Heeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeello")} // invalid
;
}
It depends if there is an other instruction after that instruction.
Runnable oneLineValid = () -> {
System.out.println("Heeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeello") // valid
;
System.out.println("Heeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeello2")
;
}
> 
> > lambda is not advised because filter is not
> 
> Please suggest logic on how to skip such cases. If that is not possible to do without false-positives, unfortunately I can not approve such Check in main repo, but we can host it at sevntu-checks project (our sandbox project)

        objBlockWithEquals
            .entrySet().stream().filter(detailASTDetailASTEntry -> { 
                return objBlockWithHashCode.remove(detailASTDetailASTEntry.getKey()) == null; // invalid
            }).forEach(detailASTDetailASTEntry -> {
                final DetailAST equalsAST = detailASTDetailASTEntry.getValue(); // valid
                log(equalsAST, MSG_KEY_HASHCODE);
});

**Comment by romani** [comment]
>eeeeello") // valid

This is conflicting with what you have in issue description.

Please move all cases to description and keep comment "violation"/"ok" on each target code.

**Comment by Dichotomia** [comment]
The intellij check https://github.com/JetBrains/intellij-community/blob/master/java/java-analysis-impl/src/com/intellij/codeInspection/RedundantLambdaCodeBlockInspection.java

**Comment by Vampire** [comment]
Imho this check should always advise to use expression lambda if it is possible.
This also will never change semantics (like replacing by method reference could).
If one does not like it in some circumstances, he should write suppression.

Some thoughts from me to the mentioned points:

> limit of line length for code.

Different projects have different preferences.
Some don't care about maximum line length.
Others do.
But maximium line length is not really an argument, as you can wrap it either way.
You can write
```java
Runnable oneLineValid = () -> {
  System.out.println("Heeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeello");
};
```
or you can write
```java
Runnable oneLineValid = () ->
  System.out.println("Heeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeello");
```
or you can write
```java
Runnable oneLineValid = () -> System.out.println("Heeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeello");
```

If you worry about people wanting the last variant but only if it is not resulting in too long line and if line too long want the first variant rather than the second, then maybe a parameter for the check could control this behavior?

> One more aspect is as soon lambda become multiline, is it not a rare demand to enforce `{` around.

You mean multi-line as in still one expression, but broken down to multiple lines?
Never heard of that style, but if it is demanded, maybe some property can control on whether multi-line expression should be expression lambda or not?

> ```java
> Runnable oneLineValid = () -> System.out.println("Heeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeello"); // ????
> ```

=> OK

> ```java
> Runnable oneLineValid = () -> {
>   System.out.println("Heeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeello")} // ????
> ; 

=> violation
I think OP missed the curly brace, besides that this example is not compilable as it misses the semicolon between the `)` and the `}`.
If you skip the curly brace it seems like the semicolon on the next line ends the `println` and the lambda is unclosed, that's probably why OP said it depends what follows.
I think if the missing semicolon would be there and OP recognizes the curly he would also have said violation.

> Do want to have a Check only for no argument lambda ? 

No, for all lambdas that consist only of one expression or one return statement.

> What is expected behavior on all possible type and forms of lambda?

If it can be replaced by an expression lambda legally, it should be a violation (except if there are parameters like multi-line or max line length)

> Please share more examples and try to think about lambdas that a part of chain calls .

Further examples also at #6926.
Also chained calls
violation:
```java
BiFunction<Object, Object, String> oneLineValid = (a, b) -> {
    return Stream.of(a, b)
            .filter(Objects::nonNull)
            .map(Object::toString)
            .collect(joining());
};
```
OK:
```java
BiFunction<Object, Object, String> oneLineValid = (a, b) ->
        Stream.of(a, b)
                .filter(Objects::nonNull)
                .map(Object::toString)
                .collect(joining());
```

> https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/EqualsHashCodeCheck.java#L163-L166

=> violation
should be
```java
        objBlockWithEquals
            .entrySet().stream().filter(detailASTDetailASTEntry -> objBlockWithHashCode.remove(detailASTDetailASTEntry.getKey()) == null)
            .forEach(detailASTDetailASTEntry -> {
```
or
```java
        objBlockWithEquals
            .entrySet().stream().filter(detailASTDetailASTEntry ->
                objBlockWithHashCode.remove(detailASTDetailASTEntry.getKey()) == null
            ).forEach(detailASTDetailASTEntry -> {
```

> lambda is not advised because filter is not supposed to modify state of objects:
> Predicate is supposed to be stateless

This should have nothing to do with stateful or stateless.
If it can be replaced by an expression lambda, it should be a violation, if not then not.
